cli-plugins/socket/socket.go

```diff
@@ -7,16 +7,98 @@ import (
 	"io"
 	"net"
 	"os"
+	"sync"
+	"time"
 )
 
 // EnvKey represents the well-known environment variable used to pass the plugin being
 // executed the socket name it should listen on to coordinate with the host CLI.
 const EnvKey = "DOCKER_CLI_PLUGIN_SOCKET"
 
+// UnixConnSafe is a wrapper around a UnixConn
+// that makes it concurrency-safe.
+type UnixConnSafe struct {
+	UnixConn   *net.UnixConn
+	SocketLock *sync.Mutex
+}
+
+var _ net.Conn = (*UnixConnSafe)(nil)
+
+func NewUnixConnSafe() *UnixConnSafe {
+	return &UnixConnSafe{
+		UnixConn:   &net.UnixConn{},
+		SocketLock: &sync.Mutex{},
+	}
+}
+
+func (c *UnixConnSafe) Close() error {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	if c.UnixConn == nil {
+		return nil
+	}
+	return c.UnixConn.Close()
+}
+
+// LocalAddr implements net.Conn.
+func (c *UnixConnSafe) LocalAddr() net.Addr {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.LocalAddr()
+}
+
+// Read implements net.Conn.
+func (c *UnixConnSafe) Read(b []byte) (n int, err error) {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.Read(b)
+}
+
+// RemoteAddr implements net.Conn.
+func (c *UnixConnSafe) RemoteAddr() net.Addr {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.RemoteAddr()
+}
+
+// SetDeadline implements net.Conn.
+func (c *UnixConnSafe) SetDeadline(t time.Time) error {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.SetDeadline(t)
+}
+
+// SetReadDeadline implements net.Conn.
+func (c *UnixConnSafe) SetReadDeadline(t time.Time) error {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.SetReadDeadline(t)
+}
+
+// SetWriteDeadline implements net.Conn.
+func (c *UnixConnSafe) SetWriteDeadline(t time.Time) error {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.SetWriteDeadline(t)
+}
+
+// Write implements net.Conn.
+func (c *UnixConnSafe) Write(b []byte) (n int, err error) {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn.Write(b)
+}
+
+func (c *UnixConnSafe) IsNil() bool {
+	c.SocketLock.Lock()
+	defer c.SocketLock.Unlock()
+	return c.UnixConn == nil
+}
+
 // SetupConn sets up a Unix socket listener, establishes a goroutine to handle connections
 // and update the conn pointer, and returns the listener for the socket (which the caller
 // is responsible for closing when it's no longer needed).
-func SetupConn(conn **net.UnixConn) (*net.UnixListener, error) {
+func SetupConn(conn *UnixConnSafe) (*net.UnixListener, error) {
```

Comment:
If we decide to go with this implementation, I think I'd prefer keeping the `UnixConnSafe`  internal. This function could continue to accept a regular `net.UnixConn` and wrap it internally.